File: library/src/main/java/com/yy/codex/uikit/UIButton.java
package com.yy.codex.uikit;

import android.content.Context;
import android.os.Build;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.annotation.RequiresApi;
import android.util.AttributeSet;
import android.view.View;

import java.util.EnumSet;
import java.util.HashMap;

/**
 * Created by cuiminghui on 2017/1/16.
 */

public class UIButton extends UIControl {

    public UIButton(@NonNull Context context, @NonNull View view) {
        super(context, view);
    }

    public UIButton(@NonNull Context context) {
        super(context);
    }

    public UIButton(@NonNull Context context, @NonNull AttributeSet attrs) {
        super(context, attrs);
    }

    public UIButton(@NonNull Context context, @NonNull AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
    public UIButton(@NonNull Context context, @NonNull AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
    }

    @Override
    protected void init() {
        super.init();
        initTitleLabel();
        initImageView();
    }

    /* UIControl */

    @Override
    protected void resetState() {
        super.resetState();
        resetTitleLabel();
        resetImageView();
    }

    /* UIView */

    @Override
    public void tintColorDidChanged() {
        super.tintColorDidChanged();
        resetTitleLabel();
    }

    /* Title */

    @NonNull private UILabel mTitleLabel;

    private void initTitleLabel() {
        mTitleLabel = new UILabel(getContext());
        mTitleLabel.setNumberOfLines(1);
        mTitleLabel.setLinebreakMode(NSLineBreakMode.ByTruncatingMiddle);
        addSubview(mTitleLabel);
    }

    private void resetTitleLabel() {
        if (mAttributedTitles.size() > 0) {
            mTitleLabel.setAttributedText(currentAttributedTitle());
        }
        else {
            NSAttributedString attributedString = new NSAttributedString(currentTitle(), new HashMap(){{
                put(NSAttributedString.NSForegroundColorAttributeName, currentTitleColor());
                put(NSAttributedString.NSFontAttributeName, mFont);
            }});
            mTitleLabel.setAttributedText(attributedString);
        }
        layoutSubviews();
    }

    private HashMap<EnumSet<State>, String> mTitleTexts = new HashMap<>();

    @NonNull
    protected String currentTitle() {
        EnumSet<State> state = getState();
        if (mTitleTexts.get(state) != null) {
            return mTitleTexts.get(state);
        }
        else if (mTitleTexts.get(EnumSet.of(State.Normal)) != null){
            return mTitleTexts.get(EnumSet.of(State.Normal));
        }
        else {
            return "";
        }
    }

    public void setTitle(@NonNull String title, State state) {
        setTitle(title, EnumSet.of(State.Normal, state));
    }

    public void setTitle(@NonNull String title, EnumSet<State> state) {
        if (state.contains(State.Selected)) {
            state.remove(State.Normal);
        }
        mTitleTexts.put(state, title);
        resetTitleLabel();
    }

    private HashMap<EnumSet<State>, UIColor> mTitleColors = new HashMap<>();

    @NonNull
    protected UIColor currentTitleColor() {
        EnumSet<State> state = getState();
        if (mTitleColors.get(state) != null) {
            return mTitleColors.get(state);
        }
        else if (mTitleColors.get(EnumSet.of(State.Normal)) != null){
            return mTitleColors.get(EnumSet.of(State.Normal));
        }
        else {
            if (state.contains(State.Disabled)) {
                return UIColor.blackColor.colorWithAlpha(0.3);
            }
            if (state.contains(State.Highlighted) && state.contains(State.Normal)) {
                return getTintColor().colorWithAlpha(0.3);
            }
            return getTintColor();
        }
    }

    public void setTitleColor(@NonNull UIColor color, State state) {
        setTitleColor(color, EnumSet.of(State.Normal, state));
    }

    public void setTitleColor(@NonNull final UIColor color, EnumSet<State> state) {
        if (state.contains(State.Selected)) {
            state.remove(State.Normal);
        }
        mTitleColors.put(state, color);
        resetTitleLabel();
    }

    private UIFont mFont = new UIFont(17);

    public void setFont(@NonNull final UIFont font) {
        mFont = font;
        resetTitleLabel();
    }

    @NonNull private HashMap<EnumSet<State>, NSAttributedString> mAttributedTitles = new HashMap<>();

    @Nullable
    public NSAttributedString currentAttributedTitle() {
        EnumSet<State> state = getState();
        if (mAttributedTitles.get(state) != null) {
            return mAttributedTitles.get(state);
        }
        else if (mAttributedTitles.get(EnumSet.of(State.Normal)) != null){
            return mAttributedTitles.get(EnumSet.of(State.Normal));
        }
        else {
            return null;
        }
    }

    public void setAttributedTitle(@NonNull NSAttributedString attributedString, State state) {
        setAttributedTitle(attributedString, EnumSet.of(State.Normal, state));
    }

    public void setAttributedTitle(@NonNull NSAttributedString attributedString, EnumSet<State> state) {
        if (state.contains(State.Selected)) {
            state.remove(State.Normal);
        }
        mAttributedTitles.put(state, attributedString);
        resetTitleLabel();
    }

    /* ImageView */

    @NonNull UIImageView mImageView;

    private void initImageView() {
        mImageView = new UIImageView(getContext());
        addSubview(mImageView);
    }

    private void resetImageView() {
        UIImage image = currentImage();
        if (image == null) {
            return;
        }
        if (image.getRenderingMode() != UIImage.RenderingMode.AlwaysOriginal) {
            mImageView.getLayer().setBitmapColor(currentTitleColor());
        }
        else {
            mImageView.getLayer().setBitmapColor(null);
        }
        mImageView.setImage(image);
        mImageView.invalidate();
        layoutSubviews();
    }

    @NonNull private HashMap<EnumSet<State>, UIImage> mImages = new HashMap<>();

    @Nullable
    public UIImage currentImage() {
        EnumSet<State> state = getState();
        if (mImages.get(state) != null) {
            return mImages.get(state);
        }
        else if (mImages.get(EnumSet.of(State.Normal)) != null){
            return mImages.get(EnumSet.of(State.Normal));
        }
        else {
            return null;
        }
    }

    public void setImage(UIImage image, State state) {
        setImage(image, EnumSet.of(state));
    }

    public void setImage(@NonNull UIImage image, EnumSet<State> state) {
        if (state.contains(State.Selected)) {
            state.remove(State.Normal);
        }
        mImages.put(state, image);
        resetImageView();
    }

    /* Layouts */

    private UIEdgeInsets mContentEdgeInsets = new UIEdgeInsets(0, 0, 0, 0);
    private UIEdgeInsets mTitleEdgeInsets = new UIEdgeInsets(0, 0, 0, 0);
    private UIEdgeInsets mImageEdgeInsets = new UIEdgeInsets(0, 0, 0, 0);

    public void setContentEdgeInsets(UIEdgeInsets contentEdgeInsets) {
        this.mContentEdgeInsets = contentEdgeInsets;
        layoutSubviews();
    }

    public void setTitleEdgeInsets(UIEdgeInsets titleEdgeInsets) {
        this.mTitleEdgeInsets = titleEdgeInsets;
        layoutSubviews();
    }

    public void setImageEdgeInsets(UIEdgeInsets imageEdgeInsets) {
        this.mImageEdgeInsets = imageEdgeInsets;
        layoutSubviews();
    }

    @Override
    public void layoutSubviews() {
        super.layoutSubviews();
        mTitleLabel.setMaxWidth(getFrame().size.width - mImageView.intrinsicContentSize().width - (mContentEdgeInsets.left + mContentEdgeInsets.right + mImageEdgeInsets.left + mImageEdgeInsets.right + mTitleEdgeInsets.left + mTitleEdgeInsets.right));
        double contentWidth = mTitleLabel.intrinsicContentSize().width + mImageView.intrinsicContentSize().width;
        contentWidth += mContentEdgeInsets.left + mContentEdgeInsets.right + mImageEdgeInsets.left + mImageEdgeInsets.right + mTitleEdgeInsets.left + mTitleEdgeInsets.right;
        double imageViewOriginY = 0;
        double titleLabelOriginY = 0;
        if (getContentVerticalAlignment() == ContentVerticalAlignment.Center) {
            imageViewOriginY = (getFrame().size.height - mImageView.intrinsicContentSize().height) / 2.0;
            titleLabelOriginY = (getFrame().size.height - mTitleLabel.intrinsicContentSize().height) / 2.0;
        }
        else if (getContentVerticalAlignment() == ContentVerticalAlignment.Bottom) {
            imageViewOriginY = getFrame().size.height - mImageView.intrinsicContentSize().height;
            titleLabelOriginY = getFrame().size.height - mTitleLabel.intrinsicContentSize().height;
        }
        if (getContentHorizontalAlignment() == ContentHorizontalAlignment.Left) {
            mImageView.setFrame(new CGRect(
                    mContentEdgeInsets.left + mImageEdgeInsets.left,
                    imageViewOriginY,
                    mImageView.intrinsicContentSize().width,
                    mImageView.intrinsicContentSize().height)
            );
            mTitleLabel.setFrame(new CGRect(
                    mContentEdgeInsets.left + mImageEdgeInsets.left + mImageView.getFrame().size.width + mImageEdgeInsets.right + mTitleEdgeInsets.left,
                    titleLabelOriginY,
                    mTitleLabel.intrinsicContentSize().width,
                    mTitleLabel.intrinsicContentSize().height)
            );
        }
        else if (getContentHorizontalAlignment() == ContentHorizontalAlignment.Center) {
            mImageView.setFrame(new CGRect(
                    ((getFrame().size.width - contentWidth) / 2.0) + mContentEdgeInsets.left + mImageEdgeInsets.left,
                    imageViewOriginY,
                    mImageView.intrinsicContentSize().width,
                    mImageView.intrinsicContentSize().height)
            );
            mTitleLabel.setFrame(new CGRect(
                    ((getFrame().size.width - contentWidth) / 2.0) + mContentEdgeInsets.left + mImageEdgeInsets.left + mImageView.getFrame().size.width + mImageEdgeInsets.right + mTitleEdgeInsets.left,
                    titleLabelOriginY,
                    mTitleLabel.intrinsicContentSize().width,
                    mTitleLabel.intrinsicContentSize().height)
            );
        }
        else if (getContentHorizontalAlignment() == ContentHorizontalAlignment.Right) {
            mImageView.setFrame(new CGRect(
                    (getFrame().size.width - contentWidth) + mContentEdgeInsets.left + mImageEdgeInsets.left,
                    imageViewOriginY,
                    mImageView.intrinsicContentSize().width,
                    mImageView.intrinsicContentSize().height)
            );
            mTitleLabel.setFrame(new CGRect(
                    (getFrame().size.width - contentWidth) + mContentEdgeInsets.left + mImageEdgeInsets.left + mImageView.getFrame().size.width + mImageEdgeInsets.right + mTitleEdgeInsets.left,
                    titleLabelOriginY,
                    mTitleLabel.intrinsicContentSize().width,
                    mTitleLabel.intrinsicContentSize().height)
            );
        }
    }
}

File: library/src/main/java/com/yy/codex/uikit/UIButton.kt
package com.yy.codex.uikit

import android.content.Context
import android.os.Build
import android.support.annotation.RequiresApi
import android.util.AttributeSet
import android.view.View
import java.util.*

/**
 * Created by cuiminghui on 2017/1/16.
 */

open class UIButton : UIControl {

    /* Props */

    lateinit var titleLabel: UILabel
        private set
    private var titles = HashMap<EnumSet<UIControl.State>, String>()
    private var titleColors = HashMap<EnumSet<UIControl.State>, UIColor>()
    private var attributedTitles = HashMap<EnumSet<UIControl.State>, NSAttributedString>()
    var font = UIFont(17f)
        set(value) {
            field = value
            resetTitleLabel()
        }
    lateinit var imageView: UIImageView
        private set
    private var images = HashMap<EnumSet<UIControl.State>, UIImage>()
    var contentEdgeInsets = UIEdgeInsets(0.0, 0.0, 0.0, 0.0)
        set(value) {
            field = value
            layoutSubviews()
        }
    var titleEdgeInsets = UIEdgeInsets(0.0, 0.0, 0.0, 0.0)
        set(value) {
            field = value
            layoutSubviews()
        }
    var imageEdgeInsets = UIEdgeInsets(0.0, 0.0, 0.0, 0.0)
        set(value) {
            field = value
            layoutSubviews()
        }

    constructor(context: Context, view: View) : super(context, view) {}

    constructor(context: Context) : super(context) {}

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {}

    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {}

    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int, defStyleRes: Int) : super(context, attrs, defStyleAttr, defStyleRes) {}

    override fun init() {
        super.init()
        initTitleLabel()
        initImageView()
    }

    override fun prepareProps(attrs: AttributeSet) {
        super.prepareProps(attrs)
        val typedArray = context.theme.obtainStyledAttributes(attrs, R.styleable.UIButton, 0, 0)
        typedArray.getString(R.styleable.UIButton_button_title)?.let {
            initializeAttributes.put("UIButton.title", it)
        }
        typedArray.getResourceId(R.styleable.UIButton_button_image, -1)?.let {
            if (it != -1) {
                initializeAttributes.put("UIButton.image", it)
            }
        }
    }

    override fun resetProps() {
        super.resetProps()
        (initializeAttributes["UIButton.title"] as? String)?.let {
            setTitle(it, State.Normal)
        }
        (initializeAttributes["UIButton.image"] as? Int)?.let {
            setImage(UIImage(context, it), State.Normal)
        }
    }

    /* UIControl */

    override fun resetState() {
        super.resetState()
        resetTitleLabel()
        resetImageView()
    }

    /* UIView */

    override fun tintColorDidChanged() {
        super.tintColorDidChanged()
        resetTitleLabel()
        resetImageView()
    }

    /* Title */

    private fun initTitleLabel() {
        titleLabel = UILabel(context)
        titleLabel.numberOfLines = 1
        titleLabel.linebreakMode = NSLineBreakMode.ByTruncatingMiddle
        addSubview(titleLabel)
    }

    private fun resetTitleLabel() {
        titleLabel?.let {
            if (attributedTitles.size > 0) {
                it.attributedText = currentAttributedTitle()
            } else {
                val attributedString = NSAttributedString(currentTitle() ?: "", hashMapOf(
                        Pair(NSAttributedString.NSForegroundColorAttributeName, currentTitleColor() ?: UIColor.blackColor),
                        Pair(NSAttributedString.NSFontAttributeName, font)
                ))
                it.attributedText = attributedString
            }
            layoutSubviews()
        }
    }

    fun currentTitle(): String? {
        val state = state
        if (titles[state] != null) {
            return titles[state]
        } else if (titles[EnumSet.of(UIControl.State.Normal)] != null) {
            return titles[EnumSet.of(UIControl.State.Normal)]
        } else {
            return ""
        }
    }

    fun setTitle(title: String, state: UIControl.State) {
        setTitle(title, EnumSet.of(UIControl.State.Normal, state))
    }

    fun setTitle(title: String, state: EnumSet<UIControl.State>) {
        if (state.contains(UIControl.State.Selected)) {
            state.remove(UIControl.State.Normal)
        }
        titles.put(state, title)
        resetTitleLabel()
    }

    fun currentTitleColor(): UIColor? {
        val state = state
        if (titleColors[state] != null) {
            return titleColors[state]
        } else if (titleColors[EnumSet.of(UIControl.State.Normal)] != null) {
            return titleColors[EnumSet.of(UIControl.State.Normal)]
        } else {
            if (state.contains(UIControl.State.Disabled)) {
                return UIColor.blackColor.colorWithAlpha(0.3)
            }
            if (state.contains(UIControl.State.Highlighted) && state.contains(UIControl.State.Normal)) {
                return tintColor?.colorWithAlpha(0.3) ?: tintColor
            }
            return tintColor
        }
    }

    fun setTitleColor(color: UIColor, state: UIControl.State) {
        setTitleColor(color, EnumSet.of(UIControl.State.Normal, state))
    }

    fun setTitleColor(color: UIColor, state: EnumSet<UIControl.State>) {
        if (state.contains(UIControl.State.Selected)) {
            state.remove(UIControl.State.Normal)
        }
        titleColors.put(state, color)
        resetTitleLabel()
    }

    fun currentAttributedTitle(): NSAttributedString? {
        val state = state
        if (attributedTitles[state] != null) {
            return attributedTitles[state]
        } else if (attributedTitles[EnumSet.of(UIControl.State.Normal)] != null) {
            return attributedTitles[EnumSet.of(UIControl.State.Normal)]
        } else {
            return null
        }
    }

    fun setAttributedTitle(attributedString: NSAttributedString, state: UIControl.State) {
        setAttributedTitle(attributedString, EnumSet.of(UIControl.State.Normal, state))
    }

    fun setAttributedTitle(attributedString: NSAttributedString, state: EnumSet<UIControl.State>) {
        if (state.contains(UIControl.State.Selected)) {
            state.remove(UIControl.State.Normal)
        }
        attributedTitles.put(state, attributedString)
        resetTitleLabel()
    }

    /* ImageView */

    private fun initImageView() {
        imageView = UIImageView(context)
        addSubview(imageView)
    }

    private fun resetImageView() {
        val image = currentImage() ?: return
        if (image.renderingMode != UIImage.RenderingMode.AlwaysOriginal) {
            imageView.layer.bitmapColor = currentTitleColor()
        } else {
            imageView.layer.bitmapColor = null
        }
        imageView.image = image
        imageView.invalidate()
        layoutSubviews()
    }



    fun currentImage(): UIImage? {
        val state = state
        if (images[state] != null) {
            return images[state]
        } else if (images[EnumSet.of(UIControl.State.Normal)] != null) {
            return images[EnumSet.of(UIControl.State.Normal)]
        } else {
            return null
        }
    }

    fun setImage(image: UIImage, state: UIControl.State) {
        setImage(image, EnumSet.of(state))
    }

    fun setImage(image: UIImage, state: EnumSet<UIControl.State>) {
        if (state.contains(UIControl.State.Selected)) {
            state.remove(UIControl.State.Normal)
        }
        images.put(state, image)
        resetImageView()
    }

    /* Layouts */

    override fun intrinsicContentSize(): CGSize {
        titleLabel.maxWidth = 999999.0
        var contentWidth = titleLabel.intrinsicContentSize().width + imageView.intrinsicContentSize().width
        contentWidth += contentEdgeInsets.left + contentEdgeInsets.right + imageEdgeInsets.left + imageEdgeInsets.right + titleEdgeInsets.left + titleEdgeInsets.right
        var contentHeight = titleLabel.intrinsicContentSize().height + imageView.intrinsicContentSize().height
        contentHeight += contentEdgeInsets.top + contentEdgeInsets.bottom + Math.max(titleEdgeInsets.top + titleEdgeInsets.bottom, imageEdgeInsets.top + imageEdgeInsets.bottom)
        return CGSize(Math.max(contentWidth, 44.0), Math.max(contentHeight, 44.0))
    }

    override fun layoutSubviews() {
        super.layoutSubviews()
        titleLabel.maxWidth = frame.size.width - imageView.intrinsicContentSize().width - (contentEdgeInsets.left + contentEdgeInsets.right + imageEdgeInsets.left + imageEdgeInsets.right + titleEdgeInsets.left + titleEdgeInsets.right)
        var contentWidth = titleLabel.intrinsicContentSize().width + imageView.intrinsicContentSize().width
        contentWidth += contentEdgeInsets.left + contentEdgeInsets.right + imageEdgeInsets.left + imageEdgeInsets.right + titleEdgeInsets.left + titleEdgeInsets.right
        var imageViewOriginY = 0.0
        var titleLabelOriginY = 0.0
        if (contentVerticalAlignment === UIControl.ContentVerticalAlignment.Center) {
            imageViewOriginY = (frame.size.height - imageView.intrinsicContentSize().height) / 2.0
            titleLabelOriginY = (frame.size.height - titleLabel.intrinsicContentSize().height) / 2.0
            imageViewOriginY = Math.ceil(imageViewOriginY)
            titleLabelOriginY = Math.ceil(titleLabelOriginY)
        } else if (contentVerticalAlignment === UIControl.ContentVerticalAlignment.Bottom) {
            imageViewOriginY = frame.size.height - imageView.intrinsicContentSize().height
            titleLabelOriginY = frame.size.height - titleLabel.intrinsicContentSize().height
            imageViewOriginY = Math.ceil(imageViewOriginY)
            titleLabelOriginY = Math.ceil(titleLabelOriginY)
        }
        if (contentHorizontalAlignment === UIControl.ContentHorizontalAlignment.Left) {
            imageView.frame = CGRect(
                    contentEdgeInsets.left + imageEdgeInsets.left,
                    imageViewOriginY,
                    imageView.intrinsicContentSize().width,
                    imageView.intrinsicContentSize().height)
            titleLabel.frame = CGRect(
                    contentEdgeInsets.left + imageEdgeInsets.left + imageView.frame.size.width + imageEdgeInsets.right + titleEdgeInsets.left,
                    titleLabelOriginY,
                    titleLabel.intrinsicContentSize().width,
                    titleLabel.intrinsicContentSize().height)
        } else if (contentHorizontalAlignment === UIControl.ContentHorizontalAlignment.Center) {
            imageView.frame = CGRect(
                    (frame.size.width - contentWidth) / 2.0 + contentEdgeInsets.left + imageEdgeInsets.left,
                    imageViewOriginY,
                    imageView.intrinsicContentSize().width,
                    imageView.intrinsicContentSize().height)
            titleLabel.frame = CGRect(
                    (frame.size.width - contentWidth) / 2.0 + contentEdgeInsets.left + imageEdgeInsets.left + imageView.frame.size.width + imageEdgeInsets.right + titleEdgeInsets.left,
                    titleLabelOriginY,
                    titleLabel.intrinsicContentSize().width,
                    titleLabel.intrinsicContentSize().height)
        } else if (contentHorizontalAlignment === UIControl.ContentHorizontalAlignment.Right) {
            imageView.frame = CGRect(
                    frame.size.width - contentWidth + contentEdgeInsets.left + imageEdgeInsets.left,
                    imageViewOriginY,
                    imageView.intrinsicContentSize().width,
                    imageView.intrinsicContentSize().height)
            titleLabel.frame = CGRect(
                    frame.size.width - contentWidth + contentEdgeInsets.left + imageEdgeInsets.left + imageView.frame.size.width + imageEdgeInsets.right + titleEdgeInsets.left,
                    titleLabelOriginY,
                    titleLabel.intrinsicContentSize().width,
                    titleLabel.intrinsicContentSize().height)
        }
    }

}
